test/performance/range/gap_decorator_benchmark.cpp

```diff
@@ -0,0 +1,460 @@
+    gaps.resize(gap_pos);      // trim sampled gap vector
+}
+
+/* Helper function to prepare a gapped sequence for the benchmark (case gap_flag=true)
+ *
+ * Parameters:
+ * size_type        size type of the sequence
+ * iterator_type    iterator type of decorated sequence iterator
+ * gap_decorator_t  gap decorator type, e.g. gap_decorator_anchor_set
+ *
+ * Arguments:
+ * gaps             reference to gap vector
+ * gap_decorator    reference to gap decorator
+ */
+template<typename size_type, typename iterator_type, typename gap_decorator_t>
+void insert_gaps(std::vector<size_type> & gaps, gap_decorator_t & gap_decorator)
+{
+    [[maybe_unused]] size_type gap_acc = 0;
+    for (size_type i = 0; i < gaps.size(); ++i)
+    {
+        if (gaps[i])
+        {
+            iterator_type it = std::ranges::begin(gap_decorator);
+            std::ranges::advance(it, std::min(i + gap_acc, gap_decorator.size()));
+            insert_gap(gap_decorator, it, gaps[i]);
+        }
+        gap_acc += gaps[i];
+    }
+}
+
+// ============================================================================
+//  read left to right (looped in case #ops exceeds sequence length)
+// ============================================================================
+/* Parameters:
+ * gap_decorator_t      gap decorator class, e.g. gap_decorator_anchor_set
+ * unaligned_seq_t     unaligned sequence to which a reference is stored
+ * alphabet_t           alphabet type of the ungapped sequence
+ * gapped_flag          operate on already gapped (true) or ungapped sequence (false)
+ */
+
+
+template <typename gap_decorator_t, bool gapped_flag>
+static void read_left2right(benchmark::State& state)
+{
+    // get target sequence length from current range state
+    unsigned int seq_len = state.range(0);
+    using size_type = typename gap_decorator_t::size_type;
+    using iterator_type = typename gap_decorator_t::iterator;
+    using sequence_type = remove_cvref_t<detail::unaligned_seq_t<gap_decorator_t>>;
+    sequence_type seq(seq_len, 'A'_dna4);
+
+    // vector of sampled gap lengths for each position
+    std::vector<size_type> gaps(seq_len, 0);
+    sample<size_type>(&gaps, seq_len);
+
+    // determine sum of gaps and non-gap symbols for not exceeding targeted sequence length
+    if constexpr(gapped_flag)
+        resize<size_type, sequence_type>(gaps, seq, seq_len);
+
+    // initialize with (truncated) sequence and insert gaps from left to right
+    gap_decorator_t gap_decorator;
+    assign_unaligned(gap_decorator, seq);
+
+    // insert gaps before starting benchmark
+    if constexpr(gapped_flag)
+        insert_gaps<size_type, iterator_type, gap_decorator_t>(gaps, gap_decorator);
+
+    size_t op_ctr = 0;
+    for (auto _ : state)
+    {
+        state.PauseTiming();
+        size_t pos = op_ctr % seq_len;
+        state.ResumeTiming();
+        benchmark::DoNotOptimize(gap_decorator[pos]);
```

Comment:
As I understand `DoNotOptimize` forces the compiler to store the result, but it is not applied to intermediate commands. So, the compilation of `gap_decorator.someOp()` is not changed to something like `-O0`, it will be compiled with whatever flags had been given (e.g. `-O3`). The `DoNotOptimize` only forces execution of the right side.